drivers/watchdog/wdt_fwdgt_gd32.c

```diff
@@ -0,0 +1,163 @@
+	 ((psc) == FWDGT_PSC_DIV128) ||	\
+	 ((psc) == FWDGT_PSC_DIV256))
+
+#if (CONFIG_FWDGT_GD32_INITIAL_TIMEOUT <= 0)
+#error Must be FWDGT_GD32_INITIAL_TIMEOUT > 0
+#elif (CONFIG_FWDGT_GD32_INITIAL_TIMEOUT > (FWDGT_PRESCALER_MAX * FWDGT_RELOAD_MAX * MSEC_PER_SEC \
+					    / CONFIG_GD32_LOW_SPEED_IRC_FREQUENCY))
+#error Must be FWDGT_GD32_INITIAL_TIMEOUT <= (256 * 4095 * 1000 / GD32_LOW_SPEED_IRC_FREQUENCY)
+#endif
+
+/**
+ * @brief Calculates FWDGT config value from timeout.
+ *
+ * @param timeout Timeout value in milliseconds.
+ * @param prescaler Pointer to the storage of prescaler value.
+ * @param reload Pointer to the storage of reload value.
+ *
+ * @return 0 on success, -EINVAL if the timeout is out of range
+ */
+static int gd32_fwdgt_convert_timeout(uint32_t timeout,
+				      uint32_t *prescaler,
+				      uint32_t *reload)
+{
+	uint16_t divider = 4U;
+	uint8_t shift = 0U;
+	uint32_t ticks = (uint64_t)CONFIG_GD32_LOW_SPEED_IRC_FREQUENCY
+			 * timeout / MSEC_PER_SEC;
+
+	while ((ticks / divider) > FWDGT_RELOAD_MAX) {
+		shift++;
+		divider = 4U << shift;
+	}
+
+	if (!IS_VALID_FWDGT_PRESCALER(PSC_PSC(shift)) || timeout == 0) {
+		return -EINVAL;
+	}
+
+	/* convert the 'shift' to prescaler value */
+	*prescaler = PSC_PSC(shift);
+	*reload = (ticks / divider) - 1U;
+
+	return 0;
+}
+
+static int gd32_fwdgt_setup(const struct device *dev, uint8_t options)
+{
+	ARG_UNUSED(dev);
+
+	if (options & WDT_OPT_PAUSE_HALTED_BY_DBG) {
+		dbg_periph_enable(DBG_FWDGT_HOLD);
+	}
+
+	if (options & WDT_OPT_PAUSE_IN_SLEEP) {
+		return -ENOTSUP;
+	}
+
+	fwdgt_enable();
+
+	return 0;
+}
+
+static int gd32_fwdgt_disable(const struct device *dev)
+{
+	/* watchdog cannot be stopped once started */
+	ARG_UNUSED(dev);
+
+	return -EPERM;
+}
+
+static int gd32_fwdgt_install_timeout(const struct device *dev,
+				      const struct wdt_timeout_cfg *config)
+{
+	uint32_t prescaler = 0U;
+	uint32_t reload = 0U;
+	ErrStatus errstat = ERROR;
+
+	/* Callback is not supported by FWDGT */
+	if (config->callback != NULL) {
+		return -ENOTSUP;
+	}
+
+	/* Calculate prescaler and reload value from timeout value */
+	if (gd32_fwdgt_convert_timeout(config->window.max, &prescaler, &reload) != 0) {
+		return -EINVAL;
+	}
+
+	/* Configure and run FWDGT */
+	fwdgt_write_enable();
+	errstat = fwdgt_config(reload, prescaler);
+	if (errstat != SUCCESS) {
+		return -EINVAL;
+	}
+	fwdgt_write_disable();
+
+	return 0;
+}
+
+static int gd32_fwdgt_feed(const struct device *dev, int channel_id)
+{
+	ARG_UNUSED(channel_id);
+	fwdgt_counter_reload();
+
+	return 0;
+}
+
+static const struct wdt_driver_api fwdgt_gd32_api = {
+	.setup = gd32_fwdgt_setup,
+	.disable = gd32_fwdgt_disable,
+	.install_timeout = gd32_fwdgt_install_timeout,
+	.feed = gd32_fwdgt_feed,
+};
+
+static int gd32_fwdgt_init(const struct device *dev)
+{
+	int ret = 0;
+
+	/* Turn on and wait stabilize system clock oscillator. */
+	rcu_osci_on(RCU_SCK);
+	while (!rcu_osci_stab_wait(RCU_SCK)) {
+	}
```

Comment:
should be handled by clock_control driver, see progress https://github.com/zephyrproject-rtos/zephyr/pull/41855

Reply:
OK, I'll try it.